driver-sync/src/main/com/mongodb/client/internal/KeyManagementService.java

```diff
@@ -17,36 +17,48 @@
 package com.mongodb.client.internal;
 
 import com.mongodb.ServerAddress;
+import com.mongodb.diagnostics.logging.Logger;
+import com.mongodb.diagnostics.logging.Loggers;
 import com.mongodb.internal.connection.SslHelper;
 
+import javax.net.SocketFactory;
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLParameters;
 import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.net.InetAddress;
 import java.net.InetSocketAddress;
 import java.net.Socket;
 import java.nio.ByteBuffer;
+import java.util.Map;
+
+import static com.mongodb.assertions.Assertions.notNull;
 
 class KeyManagementService {
-    private final SSLContext sslContext;
-    private final int defaultPort;
+    private static final Logger LOGGER = Loggers.getLogger("client");
+    private final Map<String, SSLContext> kmsProviderSslContextMap;
     private final int timeoutMillis;
 
-    KeyManagementService(final SSLContext sslContext, final int defaultPort, final int timeoutMillis) {
-        this.sslContext = sslContext;
-        this.defaultPort = defaultPort;
+    KeyManagementService(final Map<String, SSLContext> kmsProviderSslContextMap, final int timeoutMillis) {
+        this.kmsProviderSslContextMap = notNull("kmsProviderSslContextMap", kmsProviderSslContextMap);
         this.timeoutMillis = timeoutMillis;
     }
 
-    public InputStream stream(final String host, final ByteBuffer message) throws IOException {
-        ServerAddress serverAddress = host.contains(":") ? new ServerAddress(host) : new ServerAddress(host, defaultPort);
-        SSLSocket socket = (SSLSocket) sslContext.getSocketFactory().createSocket();
+    public InputStream stream(final String kmsProvider, final String host, final ByteBuffer message) throws IOException {
+        ServerAddress serverAddress = new ServerAddress(host);
+
+        LOGGER.info("Connecting to KMS server at " + serverAddress);
+        SSLContext sslContext = kmsProviderSslContextMap.get(kmsProvider);
```

Comment:
Does this risk an NPE if there is no kmsProvider mapping - looks like null is expected?

Reply:
There is a not-null check above.  The settings classes don't allow null for this.